src/nvim/api/buffer.c

```diff
@@ -636,6 +634,297 @@ ArrayOf(Integer, 2) buffer_get_mark(Buffer buffer, String name, Error *err)
+/// If there are no following marks returns [-1, -1, -1] // TODO
+/// To search untill the end of the buffer do an upper bound of -1 //TODO
+///
+/// @param buffer The buffer handle
+/// @param namespace a identifier returned previously with namespace_create
+/// @param lower [row, col] or mark_id of lower bound
+/// @param uower [row, col] or mark_id of upper bound
+/// @param[out] err Details of an error that may have occurred
+/// @return [[mark_id, row, col], ...]
+ArrayOf(Object) buffer_mark_nextrange(Buffer buffer, Integer namespace, Object lower, Object upper, Error *err)
+{
+  Array rv = ARRAY_DICT_INIT;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  ExtendedMark *l_extmark = extmark_from_id_or_pos(buffer, namespace, lower, err);
+  ExtendedMark *u_extmark = extmark_from_id_or_pos(buffer, namespace, upper, err);
+  if(!l_extmark || !u_extmark) {
+    return rv;
+  }
+  ExtmarkArray *extmarks_in_range = extmark_nextrange(buf, (uint64_t)namespace,
+                                           l_extmark->line->lnum, l_extmark->col,
+                                           u_extmark->line->lnum, u_extmark->col);
+
+  if (!extmarks_in_range) {
+    ADD(rv, INTEGER_OBJ(-9));
+    return rv;
+  }
+
+  Array mark = ARRAY_DICT_INIT;
+  ExtendedMark *extmark;
+  for (size_t i = 0; i < kv_size(*extmarks_in_range); i++) {
+    mark.size = 0;
+    mark.capacity = 0;
+    mark.items = 0;
+    extmark = kv_A(*extmarks_in_range, i);
+    ADD(mark, INTEGER_OBJ( (Integer)extmark->mark_id) );
+    ADD(mark, INTEGER_OBJ(extmark->line->lnum));
+    ADD(mark, INTEGER_OBJ(extmark->col));
+    ADD(rv, ARRAY_OBJ(mark));
+  }
+  return rv;
+}
+
+/// Setup a new namepsace for holding you marks
+///
+/// @param namespace String name for the namespace
+/// @param[out] err Details of an error that may have occurred
+/// @return integer id to be used with future mark_ calls, or 0 if name exists
+Integer buffer_mark_ns_init(String namespace, Error *err)
+{
+  uint64_t ns_id = extmark_ns_create(namespace.data);
+  if (!ns_id) {
+    api_set_error(err, Validation, _("Namespace already exists"));
+    return 0;
+  }
+  return ns_id;
+}
+
+/// Returns a list of mark namespaces in nvim
+///
+/// @param[out] err Details of an error that may have occurred
+/// @return [[string name, int id], ...]
+ArrayOf(Object) buffer_mark_ns_ids(Error *err)
+{
+  Array rv = ARRAY_DICT_INIT;
+  Array ns_array = ARRAY_DICT_INIT;
+
+  /* TODO can delete? */
+  if (!NAMESPACES) {
+    return rv;
+  }
+
+  cstr_t key;
+  uint64_t value;
+  map_foreach(NAMESPACES, key, value, {
+    ns_array.size = 0;
+    ns_array.capacity = 0;
+    ns_array.items = 0;
+    ADD(ns_array, STRING_OBJ(cstr_to_string(key)));
+    ADD(ns_array, INTEGER_OBJ(value));
+    ADD(rv, ARRAY_OBJ(ns_array));
+  });
+  return rv;
+}
+
+
+/// Returns mark info of the mark preceding the given index
+/// If there are no previous marks returns [-1, -1, -1] // TODO
```

Comment:
> Which VimL functions? In VimL 0 means "false".
`matchdelete`

@bfredl I would also like a way to say end of buffer or start of file. I was using -1 for this. Is there some constant that we already use in other places for this?